File: sandbox-python/brimble_sandbox/resources/volumes.py
"""Volume operations."""

from __future__ import annotations

from typing import Iterator, cast

from ..constants import DEFAULT_PAGE, DEFAULT_PAGE_LIMIT, MIN_VOLUME_SIZE_GB
from ..enums import VolumeType
from ..transport import HttpTransport, RequestOptions, RetryOptions
from ..types import CreateVolumeInput, Paginated, TeamScopedPagination, Volume


class VolumesResource:
    """Manage volume lifecycle and metadata."""

    def __init__(self, transport: HttpTransport) -> None:
        self._transport = transport

    def list(
        self,
        query: TeamScopedPagination | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> Paginated:
        """List your volumes with pagination."""
        payload: dict[str, object] = {
            "page": DEFAULT_PAGE,
            "limit": DEFAULT_PAGE_LIMIT,
        }
        if query:
            payload.update(query)

        options = RequestOptions(timeout_ms=timeout_ms, retry=retry)
        return cast(
            Paginated,
            self._transport.request_json(
                endpoint="/volumes",
                method="GET",
                query=payload,
                options=options,
            ),
        )

    def iterate(
        self,
        query: TeamScopedPagination | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> Iterator[Volume]:
        """Iterate volumes across pages."""
        page = query["page"] if query and "page" in query else DEFAULT_PAGE
        limit = query["limit"] if query and "limit" in query else DEFAULT_PAGE_LIMIT

        while True:
            payload = cast(TeamScopedPagination, dict(query or {}))
            payload["page"] = page
            payload["limit"] = limit
            result = self.list(payload, timeout_ms=timeout_ms, retry=retry)
            data = cast(list[Volume], result.get("data", []))

            for volume in data:
                yield volume

            total_pages = result["totalPages"]
            if page >= total_pages or len(data) == 0:
                return

            page += 1

    def create(
        self,
        input: CreateVolumeInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        idempotency_key: str | None = None,
    ) -> Volume:
        """Create a new volume. This package only accepts `type: sandbox`."""
        volume_type = input.get("type")
        if volume_type is not None and volume_type != VolumeType.SANDBOX:
            raise ValueError("Only volume type 'sandbox' is supported by this package.")

        size_gb_raw = input.get("sizeGB")
        if size_gb_raw is None:
            raise ValueError("Volume sizeGB is required.")

        size_gb = int(size_gb_raw)
        if size_gb < MIN_VOLUME_SIZE_GB:
            raise ValueError(f"Volume size must be at least {MIN_VOLUME_SIZE_GB}GB.")

        body: dict[str, object] = dict(input)
        body["type"] = VolumeType.SANDBOX

        options = RequestOptions(timeout_ms=timeout_ms, idempotency_key=idempotency_key, retry=retry)
        return cast(
            Volume,
            self._transport.request_json(
                endpoint="/volumes",
                method="POST",
                body=body,
                options=options,
            ),
        )

    def get(
        self,
        volume_id: str,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> Volume:
        """Fetch one volume by id."""
        options = RequestOptions(timeout_ms=timeout_ms, retry=retry)
        return cast(
            Volume,
            self._transport.request_json(
                endpoint=f"/volumes/{volume_id}",
                method="GET",
                options=options,
            ),
        )

    def delete(
        self,
        volume_id: str,
        *,
        timeout_ms: int | None = None,
        idempotency_key: str | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> None:
        """Delete a volume by id."""
        options = RequestOptions(timeout_ms=timeout_ms, idempotency_key=idempotency_key, retry=retry)
        self._transport.request_json(
            endpoint=f"/volumes/{volume_id}",
            method="DELETE",
            options=options,
        )

File: sandbox/src/resources/volumes.ts
import type { RequestOptions } from '../transport/http';
import { HttpTransport } from '../transport/http';
import { toPaginationQuery } from '../transport/pagination';
import { VolumeType } from '../enums';
import { DEFAULT_PAGE, DEFAULT_PAGE_LIMIT, MIN_VOLUME_SIZE_GB } from '../constants';
import type { CreateVolumeInput, Paginated, TeamScopedPagination, Volume } from '../types';

export class VolumesResource {
  private readonly transport: HttpTransport;

  /** @internal Create the volumes resource wrapper. */
  public constructor(transport: HttpTransport) {
    this.transport = transport;
  }

  /** List your volumes with pagination. */
  public list(query: TeamScopedPagination = {}, options?: RequestOptions): Promise<Paginated<Volume>> {
    const params = toPaginationQuery(query);

    if (query.teamId) {
      params.set('teamId', query.teamId);
    }

    return this.transport.requestJson<Paginated<Volume>>({
      endpoint: '/volumes',
      method: 'GET',
      query: params,
      ...options,
    }) as Promise<Paginated<Volume>>;
  }

  /** Iterate over all volumes across paginated results. */
  public async *iterate(query: TeamScopedPagination = {}, options?: RequestOptions): AsyncGenerator<Volume> {
    const limit = query.limit ?? DEFAULT_PAGE_LIMIT;
    let page = query.page ?? DEFAULT_PAGE;

    while (true) {
      const paginated = await this.list({ ...query, page, limit }, options);

      for (const volume of paginated.data) {
        yield volume;
      }

      if (page >= paginated.totalPages || paginated.data.length === 0) {
        return;
      }

      page += 1;
    }
  }

  /**
   * Create a new volume.
   * This SDK accepts only `type: "sandbox"` and defaults to it when omitted.
   */
  public create(input: CreateVolumeInput, options?: RequestOptions): Promise<Volume> {
    if (input.type && input.type !== VolumeType.Sandbox) {
      throw new Error('Only volume type "sandbox" is supported by this package.');
    }

    if (input.sizeGB < MIN_VOLUME_SIZE_GB) {
      throw new Error(`Volume size must be at least ${MIN_VOLUME_SIZE_GB}GB.`);
    }

    const body: CreateVolumeInput = {
      ...input,
      type: VolumeType.Sandbox,
    };

    return this.transport.requestJson<Volume>({
      endpoint: '/volumes',
      method: 'POST',
      body,
      ...options,
    }) as Promise<Volume>;
  }

  /** Fetch one volume by id. */
  public get(volumeId: string, options?: RequestOptions): Promise<Volume> {
    return this.transport.requestJson<Volume>({
      endpoint: `/volumes/${volumeId}`,
      method: 'GET',
      ...options,
    }) as Promise<Volume>;
  }

  /** Delete a volume by id. */
  public async delete(volumeId: string, options?: RequestOptions): Promise<void> {
    await this.transport.requestJson({
      endpoint: `/volumes/${volumeId}`,
      method: 'DELETE',
      ...options,
    });
  }
}
